Support PSA Crypto from TF-M

LDong-Arm opened the pull request:
Use only the standard PSA Crypto API, to support TF-M:
* Check PSA Crypto API major version instead of Mbed TLS version, to avoid direct dependency on an Mbed TLS header.
* Call `mbedtls_psa_crypto_free()` (not part of the PSA Crypto API) when using Mbed OS PSA only.
* Enable SHA-256 hash for PKCS1v15 message signing, because CryptoCell 312 does not allow no hash.
* Reword: Mbed Crypto -> PSA Crypto
* Add ARM_MUSCA_S1 to Travis as a TF-M target. Also update the Ubuntu image to Focal, hoping to solve the bootloader image signing failure.

General fixes & improvements
* Fix Travis build
* Update README with Mbed CLI 2 instructions
* Enable PSA by default (to make it slightly easier for users to enable Mbed Crypto)

LDong-Arm committed:
.travis.yml: Fix Travis build

* Update commands for mbed-tools which have been renamed.
* Switch to use Ubuntu Focal as Xenial has reached its End of Life.
* Install python packages using mbed-os/requirements.txt instead of
hardcoding them.

LDong-Arm committed:
Enable PSA by default

This example requires PSA but some targets do not have it enabled
by default, so enable it in the application's configuration.

Note: A user still needs to enable a PSA implementation if their
target does not have one by default.

LDong-Arm committed:
Update README with Mbed CLI 2 instructions

LDong-Arm committed:
Remove PSA Crypto freeing and refactor initialization

The function `mbedtls_psa_crypto_free()` is not part of the PSA Crypto
API and not provided by TF-M, so we avoid calling it in the
application. Consequently it does not make sense anymore to call
`psa_crypto_init()` repeatedly in all the examples in the application,
so we call it once at the start of `main()` only.

LDong-Arm committed:
Enhance PSA Crypto version check

Check the macro `PSA_CRYPTO_API_VERSION_MAJOR` from `crypto.h`
instead of `mbedtls/version.h`. The advantages are
* allowing different implementations of PSA Crypto
* not requiring mbed-mbedtls as a build dependency, which is
not pulled in by default when building a TF-M target.

LDong-Arm committed:
Enable SHA-256 hash for PKCS1v15 message signing

CryptoCell-312 used by some TF-M targets (e.g. Musca B1/S1) do not
support signing PKCS1v15 messages without a hash algorithm. Enabling
SHA-256 for PKCS1v15 signing fixes the `sign_a_message_using_rsa`
example on those targets.

For better readability of the example, sign an actual SHA-256 value
instead of a plain text.

LDong-Arm committed:
.travis.yml: Add ARM_MUSCA_S1 as a TF-M target

This example should be compatible with any PSA Crypto implementation,
and the existing K64F target in `.travis.yml` uses Crypto from Mbed
OS PSA. This commit adds ARM_MUSCA_S1 to ensure PSA Crypto from TF-M
also works with the example.

LDong-Arm committed:
Rewording: Mbed Crypto -> PSA Crypto

This example should work with any PSA Crypto implementations including
the one provided by TF-M, not just Mbed Crypto.

LDong-Arm commented:
Now Travis passes and this PR is ready for review!

Patater commented on getting-started/mbed_app.json:
```diff
@@ -2,7 +2,7 @@
         "*": {
              "platform.stdio-convert-newlines": true,
-             "target.features_add" : ["EXPERIMENTAL_API"],
+             "target.features_add" : ["EXPERIMENTAL_API", "PSA"],
```
Commit message in "Enable PSA by default" is doesn't seem to match what we are doing in that commit.

LDong-Arm replied:
Updated commit message.

Patater commented on getting-started/main.cpp:
```diff
@@ -161,7 +162,7 @@ static void sign_a_message_using_rsa(const uint8_t *key, size_t key_len)
 
     /* Sign message using the key */
-    status = psa_asymmetric_sign(handle, PSA_ALG_RSA_PKCS1V15_SIGN_RAW,
+    status = psa_asymmetric_sign(handle, alg,
```
Is this signing random data on the stack? Shouldn't we provide some example data to be hashed and signed? We should also check if we need to do the hash ourselves first, or if psa_asymmetric_sign() will do the hashing for us; I suspect it does the hashing for us.

LDong-Arm replied:
Very good catch. I just checked the API doc, `hash` is the _input_ and `signature` is the output, and `alg` is the signature algorithm for signing. So this indeed signs random data on the stack...

In principle, `hash` can be any value or even a plain text message. But as a demo, it's nice to actually hash some example data then sign it. :)

I'll rework this part.

LDong-Arm replied:
Oh we already have a `hash_a_message()` example. To avoid duplication I won't do hashing again, but will add a comment explaining this.

Patater replied:
If we set up a new pre-hashed buffer, that'd probably be good. SHA-256 hash of some fun message or whatever.

LDong-Arm replied:
But we won't verify the message here, because we have another `verify_a_hash()` example for that. So if we declare a message and its pre-calculated hash here, the message itself will be unused.

Patater replied:
We just need something that realistically looks like a hash for example purposes, but don't need to verify it.

LDong-Arm replied:
Done

LDong-Arm reviewed (commented):
@Patater Thanks for the review, I've addressed your comments, please have another look

Patater commented on getting-started/main.cpp:
```diff
@@ -143,32 +133,29 @@ static void import_a_key(const uint8_t *key, size_t key_len)
-    uint8_t signature[PSA_ASYMMETRIC_SIGNATURE_MAX_SIZE] = {0};
+    psa_algorithm_t alg = PSA_ALG_RSA_PKCS1V15_SIGN(PSA_ALG_SHA_256);
+    // SHA-256 of {'a', 'b', 'c'}
+    const uint8_t hash[] = {
```
I'd recommend keeping the size parameter here, to make the code a bit more robust (if a few elements got removed on accident, for example, we'd not read uninitialzed memory). I'd also recommend "static" to ensure we don't use stack memory too much.

`static const uint8_t hash[PSA_HASH_SIZE(PSA_ALG_SHA_256)]`

LDong-Arm replied:
This would be a general improvement that also applies to other examples in this application. I can add an extra commit for this.

LDong-Arm replied:
(Most others already have `static`, "verify_a_hash" where I copied the hash just happened to not have it 😮 )

Patater commented on getting-started/main.cpp:
```diff
@@ -139,8 +139,14 @@ static void sign_a_message_using_rsa(const uint8_t *key, size_t key_len)
+        0x5d, 0xae, 0x22, 0x23, 0xb0, 0x03, 0x61, 0xa3, 0x96, 0x17, 0x7a, 0x9c,
+        0xb4, 0x10, 0xff, 0x61, 0xf2, 0x00, 0x15, 0xad
+    };
+    uint8_t signature[PSA_ASYMMETRIC_SIGNATURE_MAX_SIZE];
```
Looks like the `= {0};` got dropped off here. This means there may be extra garbage at the end of the buffer after the sign() call, rather than clean 0 when you hexdump the signature buffer.

LDong-Arm replied:
Done

Patater commented on getting-started/main.cpp:
```diff
@@ -187,7 +193,7 @@ static void encrypt_with_symmetric_ciphers(const uint8_t *key, size_t key_len)
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
     psa_algorithm_t alg = PSA_ALG_CBC_NO_PADDING;
-    uint8_t plaintext[block_size] = SOME_PLAINTEXT;
+    static const uint8_t plaintext[block_size] = SOME_PLAINTEXT;
```
I'd recommend adding this static in a separate commit, for clarity and easier rework. It's not a change related to enabling "SHA-256 hash for PKCS1v15 message signing"

LDong-Arm replied:
Done

LDong-Arm committed:
Use `static const` for constant byte arrays

This reduces stack memory usage, as compilers usually put static
constant arrays in the read-only memory.

Patater reviewed (approved):
LGTM
